.github/scripts/rust/setup.py

```diff
@@ -0,0 +1,156 @@
+from fnmatch import fnmatch
+from pathlib import Path
+import re
+import json
+
+CWD = Path.cwd()
+
+# All jobs for all crates will run if any of these paths change
+ALWAYS_RUN_PATTERNS = ["**/rust-toolchain.toml", ".github/**"]
+
+# Crates which will be tested in release mode
+TEST_IN_RELEASE_CRATES = ["packages/engine"]
+
+# Exclude the stable channel for these crates
+DISABLE_STABLE_PATTERNS = ["packages/engine**"]
+
+# Try and publish these crates when their version is changed in Cargo.toml
+PUBLISH_PATTERNS = ["packages/libs/error-stack"]
+
+
+def generate_diffs():
+    """
+    Generates a diff between `HEAD^` and `HEAD`
+    """
+    from pygit2 import Repository, Commit
+
+    repository = Repository(CWD)
+    head = repository.head.peel(Commit)
+    return repository.diff(head.parents[0], head, context_lines=0)
+
+
+def search_available_crates():
+    """
+    Returns all available crates in the workspace
+    :return: a list of crate paths
+    """
+    return [path.relative_to(CWD).parent for path in CWD.rglob("Cargo.toml")]
+
+
+def filter_for_changed_crates(diffs, crates):
+    """
+    Returns a list of paths to crates which have changed files
+
+    If a file was changed, which matches `ALWAYS_RUN_PATTERNS`, all crates will be returned
+    :param diffs: a list `Diff`s returned from git
+    :param crates: a list of paths to crates
+    :return: a list of crate paths
+    """
+    # Check if any changed file matches `ALWAYS_RUN_PATTERNS`
+    if [diff for diff in diffs for pattern in ALWAYS_RUN_PATTERNS if fnmatch(diff.delta.new_file.path, pattern)]:
```

Comment:
Taking this as an opportunity to talk about good Python patterns.

```suggestion
    if any(fnmatch(diff.delta.new_file.path, pattern) for diff in diffs for pattern in ALWAYS_RUN_PATTERNS)
```

Pretty sure that should work and should short circuit on a true